File: apps/ios/UNESWidgets/WidgetSnapshot.swift
import Foundation
import SwiftUI

/// Wire-format snapshot the host app writes to the App Group container and
/// the widget extension reads back. Two targets carry their own Codable
/// type with the same field names — the JSON is the contract. Keep this in
/// sync with `apps/ios/UNES/Features/Widgets/WidgetSnapshotWriter.swift`.
///
/// Only raw schedule data lives here. Time-derived state (running /
/// upcoming / dayDone, countdowns) is recomputed from `Date()` at every
/// timeline tick, so a stale snapshot still produces a correct entry —
/// the widget just sees the same classes positioned around a newer clock.
struct WidgetSnapshot: Codable, Sendable {
    static let appGroup = "group.dev.forcetower.unes.ios"
    static let fileName = "next-class-snapshot.json"

    let generatedAt: Date
    let todayDateIso: String
    let today: [Class]
    /// First class on the next day that has any classes scheduled. Drives
    /// the `dayDone` copy. Null on Fridays / over breaks when nothing is
    /// upcoming in the week we have data for.
    let nextDay: NextDay?

    struct Class: Codable, Sendable {
        let classId: String
        let code: String
        let title: String
        let prof: String?
        let room: String?
        let topic: String?
        /// "HH:mm" — start of the class allocation.
        let startTime: String
        /// "HH:mm" — null when upstream didn't ship one.
        let endTime: String?
    }

    struct NextDay: Codable, Sendable {
        /// "YYYY-MM-DD".
        let dateIso: String
        /// Distance from today's date in days (1 = tomorrow, 2 = day after, ...).
        let daysAway: Int
        let first: Class
    }

    static func containerURL() -> URL? {
        FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroup)
    }

    static func load() -> WidgetSnapshot? {
        guard let dir = containerURL() else { return nil }
        let url = dir.appendingPathComponent(fileName)
        guard let data = try? Data(contentsOf: url) else { return nil }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try? decoder.decode(WidgetSnapshot.self, from: data)
    }

    /// Returns the effective "now" the provider should render against. In
    /// DEBUG builds, an ISO8601 string under
    /// `UserDefaults(suiteName: appGroup)["widget.debug.now"]` overrides the
    /// system clock so we can preview widget states without changing the
    /// host Mac clock. Set via:
    ///
    ///   xcrun simctl spawn booted defaults write group.dev.forcetower.unes.ios \
    ///     widget.debug.now -string "2026-05-12T10:00:00-03:00"
    ///
    /// Clear via `defaults delete`. Compiles out in Release.
    static func resolvedNow() -> Date {
        #if DEBUG
        if let override = debugClockOverride() {
            return override
        }
        #endif
        return Date()
    }

    /// Same idea as `resolvedNow()` but seeded from a `TimelineView` context.
    /// In production this just hands back `contextDate` so `.everyMinute`
    /// ticks drive the live countdowns. In DEBUG with the clock override
    /// set we ignore the tick and return the override so the view stays
    /// frozen at the requested moment for visual testing.
    static func effectiveNow(contextDate: Date) -> Date {
        #if DEBUG
        if let override = debugClockOverride() {
            return override
        }
        #endif
        return contextDate
    }

    #if DEBUG
    static var isDebugClockOverridden: Bool {
        debugClockOverride() != nil
    }

    private static func debugClockOverride() -> Date? {
        guard let defaults = UserDefaults(suiteName: appGroup),
              let raw = defaults.string(forKey: "widget.debug.now") else { return nil }
        return iso8601Parser.date(from: raw)
    }

    private static let iso8601Parser: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()
    #endif
}

// MARK: - Entry rendering

extension WidgetSnapshot {
    /// Pure derivation of a `NextClassEntry` from this snapshot at `now`.
    /// State + countdowns recomputed from absolute times — the host doesn't
    /// have to rewrite the file just because the clock advanced.
    func renderEntry(at now: Date, calendar: Calendar = .current) -> NextClassEntry {
        let nowMinutes = minutesOfDay(now, in: calendar)
        let todayIsToday = todayDateIso == isoDate(now, in: calendar)

        // If the snapshot was written on a different local date than now,
        // today's allocations no longer apply — fall through to nextDay if
        // present, else dayDone.
        let currentClasses = todayIsToday ? today : []

        let running = currentClasses.first { c in
            guard let start = parseHhMm(c.startTime), let end = parseHhMm(c.endTime) else { return false }
            return nowMinutes >= start && nowMinutes < end
        }

        let upcoming = currentClasses
            .compactMap { c -> (Class, Int)? in
                guard let start = parseHhMm(c.startTime), start > nowMinutes else { return nil }
                return (c, start)
            }
            .min { $0.1 < $1.1 }
            .map { $0.0 }

        let completedCount = currentClasses.filter { c in
            guard let end = parseHhMm(c.endTime) else { return false }
            return nowMinutes >= end
        }.count

        let bars = currentClasses.map { c -> NextClassEntry.TodayBar in
            let end = parseHhMm(c.endTime) ?? parseHhMm(c.startTime) ?? 0
            let state: NextClassEntry.TodayBar.State
            if nowMinutes >= end {
                state = .done
            } else if let next = upcoming, next.classId == c.classId {
                state = .next
            } else if let r = running, r.classId == c.classId {
                state = .next
            } else {
                state = .later
            }
            return NextClassEntry.TodayBar(
                code: shortCode(from: c.code),
                time: c.startTime,
                state: state,
                colorHex: WidgetColor.subjectHex(for: c.code)
            )
        }

        // In the final stretch of a running class we shift attention to the
        // next class so the student gets a heads-up before walking out, but
        // only when there's actually another class to flag — otherwise we
        // keep showing the running one so the widget doesn't go blank for
        // the last 30 minutes of the day.
        let inClassSwitchThreshold = 30
        if let r = running {
            let start = parseHhMm(r.startTime) ?? nowMinutes
            let end = parseHhMm(r.endTime) ?? (nowMinutes + 60)
            let endsIn = max(0, end - nowMinutes)
            let shouldHandOffToUpcoming = endsIn <= inClassSwitchThreshold && upcoming != nil
            if !shouldHandOffToUpcoming {
                let startDate = absoluteDate(forMinutes: start, on: now, calendar: calendar)
                let endDate = absoluteDate(forMinutes: end, on: now, calendar: calendar)
                return NextClassEntry(
                    date: now,
                    state: .inClass,
                    code: r.code,
                    shortCode: shortCode(from: r.code),
                    title: r.title,
                    shortTitle: shortTitle(from: r.title),
                    prof: r.prof ?? "",
                    room: r.room ?? "",
                    building: "",
                    startsIn: 0,
                    endsIn: endsIn,
                    totalDurationMin: max(1, end - start),
                    startTime: r.startTime,
                    endTime: r.endTime ?? "",
                    topic: r.topic,
                    referenceStart: startDate,
                    referenceEnd: endDate,
                    todayBars: bars,
                    dayDoneLine: nil,
                    completedTodayCount: completedCount
                )
            }
        }

        if let n = upcoming {
            let start = parseHhMm(n.startTime) ?? nowMinutes
            let end = parseHhMm(n.endTime) ?? start
            let startDate = absoluteDate(forMinutes: start, on: now, calendar: calendar)
            let endDate = absoluteDate(forMinutes: end, on: now, calendar: calendar)
            return NextClassEntry(
                date: now,
                state: .upcoming,
                code: n.code,
                shortCode: shortCode(from: n.code),
                title: n.title,
                shortTitle: shortTitle(from: n.title),
                prof: n.prof ?? "",
                room: n.room ?? "",
                building: "",
                startsIn: max(0, start - nowMinutes),
                endsIn: 0,
                totalDurationMin: max(1, end - start),
                startTime: n.startTime,
                endTime: n.endTime ?? "",
                topic: n.topic,
                referenceStart: startDate,
                referenceEnd: endDate,
                todayBars: bars,
                dayDoneLine: nil,
                completedTodayCount: completedCount
            )
        }

        // No class left in `currentClasses`. If the snapshot's `nextDay`
        // points at the next populated day, surface it as upcoming so the
        // main subject area on Small / Large renders real data — countdown
        // spans the day boundary. The Medium widget keeps its dedicated
        // dayDone treatment via the `dayDoneLine` carried alongside.
        //
        // Re-derive `daysAway` against render time: `n.daysAway` was
        // computed when the snapshot was written, so it drifts if the
        // snapshot is a few hours stale (production day-rollover before the
        // host writes a fresh snapshot) or in widget visual testing with a
        // debug clock override. When the re-derivation lands at 0 — i.e.
        // the next populated day is *today* — emit `.upcoming` instead of
        // `.dayDone` so the copy doesn't read "Sem aulas até hoje, 07:30"
        // while a class is genuinely coming up later today.
        if let n = nextDay {
            let nextStart = parseHhMm(n.first.startTime) ?? 0
            let nextEnd = parseHhMm(n.first.endTime) ?? nextStart
            let daysAway = max(0, calendarDaysBetween(now, isoDay: n.dateIso, calendar: calendar) ?? n.daysAway)
            let startsIn = max(0, daysAway * 1440 - nowMinutes + nextStart)
            let startDate = absoluteDate(forMinutes: nextStart, on: now, daysAhead: daysAway, calendar: calendar)
            let endDate = absoluteDate(forMinutes: nextEnd, on: now, daysAhead: daysAway, calendar: calendar)
            let isToday = daysAway == 0
            return NextClassEntry(
                date: now,
                state: isToday ? .upcoming : .dayDone,
                code: n.first.code,
                shortCode: shortCode(from: n.first.code),
                title: n.first.title,
                shortTitle: shortTitle(from: n.first.title),
                prof: n.first.prof ?? "",
                room: n.first.room ?? "",
                building: "",
                startsIn: startsIn,
                endsIn: 0,
                totalDurationMin: max(1, nextEnd - nextStart),
                startTime: n.first.startTime,
                endTime: n.first.endTime ?? "",
                topic: n.first.topic,
                referenceStart: startDate,
                referenceEnd: endDate,
                todayBars: bars,
                dayDoneLine: isToday ? nil : dayDoneLine(at: now, calendar: calendar),
                completedTodayCount: completedCount
            )
        }

        return NextClassEntry(
            date: now,
            state: .dayDone,
            code: "", shortCode: "",
            title: "", shortTitle: "",
            prof: "", room: "", building: "",
            startsIn: 0, endsIn: 0, totalDurationMin: 0,
            startTime: "", endTime: "",
            topic: nil,
            referenceStart: nil,
            referenceEnd: nil,
            todayBars: bars,
            dayDoneLine: nil,
            completedTodayCount: completedCount
        )
    }

    /// Combines `now`'s calendar day with `forMinutes` (minutes-of-day) and
    /// an optional day offset to produce an absolute Date. Used to derive
    /// `referenceStart` / `referenceEnd` from snapshot HH:MM values for the
    /// live countdown views.
    private func absoluteDate(
        forMinutes minutes: Int,
        on now: Date,
        daysAhead: Int = 0,
        calendar: Calendar
    ) -> Date? {
        let baseDay = calendar.startOfDay(for: now)
        guard let dayDate = calendar.date(byAdding: .day, value: daysAhead, to: baseDay) else { return nil }
        return calendar.date(byAdding: .minute, value: minutes, to: dayDate)
    }

    /// "amanhã, 07:30 · Cálculo I-E — PAT59" when the next class is the
    /// calendar day after `now`; otherwise substitutes the weekday name in
    /// pt-BR ("segunda, 07:30 …"). `nextDay.daysAway` was computed when the
    /// snapshot was written, so we re-derive against render time here to
    /// stay correct if the snapshot becomes a few hours stale.
    private func dayDoneLine(at now: Date, calendar: Calendar) -> String? {
        guard let nextDay else { return nil }
        let diff = calendarDaysBetween(now, isoDay: nextDay.dateIso, calendar: calendar) ?? nextDay.daysAway
        let when: String
        switch diff {
        case 1: when = "amanhã"
        case 0: when = "hoje"
        case 2...:
            when = weekdayLabel(isoDay: nextDay.dateIso) ?? "em \(diff) dias"
        default:
            when = "em breve"
        }
        var line = "\(when), \(nextDay.first.startTime) · \(nextDay.first.title)"
        if let room = nextDay.first.room, !room.isEmpty {
            line += " — \(room)"
        }
        return line
    }

    private func calendarDaysBetween(_ from: Date, isoDay: String, calendar: Calendar) -> Int? {
        guard let target = isoDayFormatter.date(from: isoDay) else { return nil }
        let start = calendar.startOfDay(for: from)
        let end = calendar.startOfDay(for: target)
        return calendar.dateComponents([.day], from: start, to: end).day
    }

    private func weekdayLabel(isoDay: String) -> String? {
        guard let date = isoDayFormatter.date(from: isoDay) else { return nil }
        return Self.weekdayFormatter.string(from: date)
            .replacingOccurrences(of: "-feira", with: "")
            .lowercased()
    }

    private static let weekdayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "pt_BR")
        f.dateFormat = "EEEE"
        f.timeZone = TimeZone.current
        return f
    }()
}

private func parseHhMm(_ value: String?) -> Int? {
    guard let value, value.count >= 4 else { return nil }
    let parts = value.split(separator: ":")
    guard parts.count >= 2, let h = Int(parts[0]), let m = Int(parts[1]) else { return nil }
    return h * 60 + m
}

private func minutesOfDay(_ date: Date, in calendar: Calendar) -> Int {
    let comps = calendar.dateComponents([.hour, .minute], from: date)
    return (comps.hour ?? 0) * 60 + (comps.minute ?? 0)
}

private let isoDayFormatter: DateFormatter = {
    let f = DateFormatter()
    f.locale = Locale(identifier: "en_US_POSIX")
    f.dateFormat = "yyyy-MM-dd"
    f.timeZone = TimeZone.current
    return f
}()

private func isoDate(_ date: Date, in calendar: Calendar) -> String {
    isoDayFormatter.timeZone = calendar.timeZone
    return isoDayFormatter.string(from: date)
}

/// Approximation of the design's `shortCode` — upstream doesn't carry one,
/// so we take the first token of the disciplina code (e.g. "MAT01023" stays
/// "MAT01023"; "CALC II" → "CALC"). Mirrors the fixture's intent.
private func shortCode(from code: String) -> String {
    let firstWord = code.split(separator: " ").first.map(String.init) ?? code
    return String(firstWord.prefix(6))
}

/// First clause before " — " / " – " / ":" if present, capped at 14 chars.
/// Upstream titles are full discipline names; the design shows a compact form.
private func shortTitle(from title: String) -> String {
    let separators = [" — ", " – ", " - ", ": "]
    var base = title
    for sep in separators {
        if let range = base.range(of: sep) {
            base = String(base[..<range.lowerBound])
        }
    }
    if base.count <= 14 { return base }
    return String(base.prefix(13)) + "…"
}

File: apps/android/app/src/main/kotlin/dev/forcetower/unes/widgets/WidgetSnapshot.kt
package dev.forcetower.unes.widgets

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File
import java.util.Calendar
import java.util.Locale
import java.util.TimeZone

// Wire-format snapshot the host process writes to internal storage and the
// widget reads back at render time. Same idea as iOS, except both producer
// and consumer live in the same APK so we share the Kotlin type directly
// instead of mirroring two struct definitions.
//
// Only raw schedule data lives here. Time-derived state (running / upcoming /
// dayDone, countdowns, today bars) is recomputed from `now` at every widget
// update via `renderEntry`, so a stale snapshot still produces a correct
// entry — the widget just sees the same classes positioned around a newer
// clock.
@Serializable
internal data class WidgetSnapshot(
    val generatedAtEpochMs: Long,
    val todayDateIso: String,
    val today: List<Class>,
    val nextDay: NextDay?,
) {
    @Serializable
    data class Class(
        val classId: String,
        val code: String,
        val title: String,
        val prof: String?,
        val room: String?,
        val topic: String?,
        // "HH:mm" — start of the class allocation.
        val startTime: String,
        // "HH:mm" — null when upstream didn't ship one.
        val endTime: String?,
    )

    @Serializable
    data class NextDay(
        // "YYYY-MM-DD".
        val dateIso: String,
        // Distance from today's date in days (1 = tomorrow, 2 = day after, ...).
        val daysAway: Int,
        val first: Class,
    )

    companion object {
        const val FILE_NAME = "next-class-snapshot.json"

        // Format used by both reader and writer; lenient so an extra field
        // shipped by a newer producer doesn't break an older consumer (and
        // vice versa during in-flight upgrades).
        private val json = Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
        }

        fun file(context: Context): File = File(context.filesDir, FILE_NAME)

        fun load(context: Context): WidgetSnapshot? {
            val f = file(context)
            if (!f.exists()) return null
            return runCatching {
                json.decodeFromString(serializer(), f.readText())
            }.getOrNull()
        }

        fun save(context: Context, snapshot: WidgetSnapshot) {
            val f = file(context)
            // Atomic replace: write a sibling tmp file then rename so a
            // half-written JSON never lands on disk for the widget reader to
            // trip over. Same pattern the iOS publisher uses with
            // `Data.write(to:options:.atomic)`.
            val tmp = File(f.parentFile, "${f.name}.tmp")
            tmp.writeText(json.encodeToString(serializer(), snapshot))
            if (!tmp.renameTo(f)) {
                // renameTo can fail across some filesystems — fall back to a
                // copy + delete which is still atomic w.r.t. partial writes
                // because the temp file contains the full payload.
                f.writeText(tmp.readText())
                tmp.delete()
            }
        }
    }
}

// Loads the on-disk snapshot and resolves it against the current wall clock
// into a `NextClassEntry`. Returns null when no snapshot has been written yet
// — caller decides whether to render a placeholder (the Glance widget) or
// fall back to a safe default cadence (the receiver). Shared by the widget
// composable and the receiver's tick scheduler so both branch on the exact
// same materialized entry.
//
// `isDark` flows into `SubjectPalette` so the discipline color baked into
// the entry uses the lifted dark-mode hex when the host renders against a
// dark UI mode. The receiver's tick path is theme-agnostic and can pass
// false (the bar/subject colors aren't read by the alarm scheduler).
internal fun loadCurrentEntry(context: Context, isDark: Boolean = false): NextClassEntry? {
    val snapshot = WidgetSnapshot.load(context) ?: return null
    val tz = TimeZone.getDefault()
    val now = Calendar.getInstance(tz)
    val nowMinutes = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE)
    val nowDateIso = String.format(
        Locale.US,
        "%04d-%02d-%02d",
        now.get(Calendar.YEAR),
        now.get(Calendar.MONTH) + 1,
        now.get(Calendar.DAY_OF_MONTH),
    )
    return snapshot.renderEntry(
        nowDateIso = nowDateIso,
        nowMinutes = nowMinutes,
        weekdayResolver = ::weekdayLabelForIsoDay,
        isDark = isDark,
    )
}

// Pure derivation of a `NextClassEntry` from this snapshot at `nowMinutes`
// + `nowDateIso`. Mirrors `WidgetSnapshot.renderEntry` on iOS — same
// running / upcoming / dayDone branch logic, same in-class-handoff threshold,
// same fall-through to `nextDay` so the Small / Large widgets render real
// data even when today is empty.
internal fun WidgetSnapshot.renderEntry(
    nowDateIso: String,
    nowMinutes: Int,
    weekdayResolver: (isoDay: String) -> String? = { null },
    isDark: Boolean = false,
): NextClassEntry {
    val todayIsToday = todayDateIso == nowDateIso
    val currentClasses = if (todayIsToday) today else emptyList()

    val running = currentClasses.firstOrNull { c ->
        val start = parseHhMm(c.startTime) ?: return@firstOrNull false
        val end = parseHhMm(c.endTime) ?: return@firstOrNull false
        nowMinutes in start until end
    }

    val upcoming = currentClasses
        .mapNotNull { c ->
            val start = parseHhMm(c.startTime) ?: return@mapNotNull null
            if (start <= nowMinutes) null else c to start
        }
        .minByOrNull { it.second }
        ?.first

    val completedCount = currentClasses.count { c ->
        val end = parseHhMm(c.endTime) ?: return@count false
        nowMinutes >= end
    }

    val bars = currentClasses.map { c ->
        val start = parseHhMm(c.startTime) ?: 0
        val end = parseHhMm(c.endTime) ?: start
        val state = when {
            nowMinutes >= end -> NextClassEntry.TodayBar.State.Done
            upcoming != null && upcoming.classId == c.classId -> NextClassEntry.TodayBar.State.Next
            running != null && running.classId == c.classId -> NextClassEntry.TodayBar.State.Next
            else -> NextClassEntry.TodayBar.State.Later
        }
        NextClassEntry.TodayBar(
            code = shortCode(c.code),
            time = c.startTime,
            state = state,
            colorArgb = SubjectPalette.argb(c.code, isDark),
        )
    }

    // In the final stretch of a running class we shift attention to the
    // next class so the student gets a heads-up before walking out, but only
    // when there's actually another class to flag. Same 30 min threshold as iOS.
    val inClassSwitchThreshold = 30
    if (running != null) {
        val start = parseHhMm(running.startTime) ?: nowMinutes
        val end = parseHhMm(running.endTime) ?: (nowMinutes + 60)
        val endsIn = (end - nowMinutes).coerceAtLeast(0)
        val handOff = endsIn <= inClassSwitchThreshold && upcoming != null
        if (!handOff) {
            return NextClassEntry(
                state = NextClassState.InClass,
                code = running.code,
                shortCode = shortCode(running.code),
                title = running.title,
                shortTitle = shortTitle(running.title),
                prof = running.prof.orEmpty(),
                room = running.room.orEmpty(),
                startsIn = 0,
                endsIn = endsIn,
                totalDurationMin = (end - start).coerceAtLeast(1),
                startTime = running.startTime,
                endTime = running.endTime.orEmpty(),
                topic = running.topic,
                todayBars = bars,
                dayDoneLine = null,
                completedTodayCount = completedCount,
                subjectColorArgb = SubjectPalette.argb(running.code, isDark),
            )
        }
    }

    if (upcoming != null) {
        val start = parseHhMm(upcoming.startTime) ?: nowMinutes
        val end = parseHhMm(upcoming.endTime) ?: start
        return NextClassEntry(
            state = NextClassState.Upcoming,
            code = upcoming.code,
            shortCode = shortCode(upcoming.code),
            title = upcoming.title,
            shortTitle = shortTitle(upcoming.title),
            prof = upcoming.prof.orEmpty(),
            room = upcoming.room.orEmpty(),
            startsIn = (start - nowMinutes).coerceAtLeast(0),
            endsIn = 0,
            totalDurationMin = (end - start).coerceAtLeast(1),
            startTime = upcoming.startTime,
            endTime = upcoming.endTime.orEmpty(),
            topic = upcoming.topic,
            todayBars = bars,
            dayDoneLine = null,
            completedTodayCount = completedCount,
            subjectColorArgb = SubjectPalette.argb(upcoming.code, isDark),
        )
    }

    // No class left in `currentClasses`. If the snapshot's `nextDay` points
    // at the next populated day, surface it as upcoming so the main subject
    // area on Small / Large renders real data — countdown spans the day
    // boundary. The Medium widget keeps its dedicated dayDone treatment via
    // `dayDoneLine`.
    //
    // Re-derive `daysAway` against render time: `n.daysAway` was computed
    // when the snapshot was written, so it drifts on day rollover before
    // the host writes a fresh snapshot. When the re-derivation lands at
    // 0 — i.e. the next populated day is *today* — emit `Upcoming` instead
    // of `DayDone` so the copy doesn't read "Sem aulas até hoje, 07:30"
    // while a class is genuinely coming up later today.
    val n = nextDay
    if (n != null) {
        val nextStart = parseHhMm(n.first.startTime) ?: 0
        val nextEnd = parseHhMm(n.first.endTime) ?: nextStart
        val daysAway = (daysBetweenIsoDays(nowDateIso, n.dateIso) ?: n.daysAway).coerceAtLeast(0)
        val startsIn = (daysAway * 1440 - nowMinutes + nextStart).coerceAtLeast(0)
        val isToday = daysAway == 0
        return NextClassEntry(
            state = if (isToday) NextClassState.Upcoming else NextClassState.DayDone,
            code = n.first.code,
            shortCode = shortCode(n.first.code),
            title = n.first.title,
            shortTitle = shortTitle(n.first.title),
            prof = n.first.prof.orEmpty(),
            room = n.first.room.orEmpty(),
            startsIn = startsIn,
            endsIn = 0,
            totalDurationMin = (nextEnd - nextStart).coerceAtLeast(1),
            startTime = n.first.startTime,
            endTime = n.first.endTime.orEmpty(),
            topic = n.first.topic,
            todayBars = bars,
            dayDoneLine = if (isToday) null else formatDayDoneLine(n, daysAway, weekdayResolver),
            completedTodayCount = completedCount,
            subjectColorArgb = SubjectPalette.argb(n.first.code, isDark),
        )
    }

    return NextClassEntry(
        state = NextClassState.DayDone,
        code = "",
        shortCode = "",
        title = "",
        shortTitle = "",
        prof = "",
        room = "",
        startsIn = 0,
        endsIn = 0,
        totalDurationMin = 0,
        startTime = "",
        endTime = "",
        topic = null,
        todayBars = bars,
        dayDoneLine = null,
        completedTodayCount = completedCount,
        subjectColorArgb = 0xFF3B9EAE.toInt(),
    )
}

// "amanhã, 07:30 · Cálculo I-E — PAT59" when the next class is the
// calendar day after now; otherwise substitutes the weekday name in pt-BR
// ("segunda, 07:30 …"). Mirrors `dayDoneLine` on iOS.
private fun formatDayDoneLine(
    nextDay: WidgetSnapshot.NextDay,
    daysAway: Int,
    weekdayResolver: (isoDay: String) -> String?,
): String {
    val when_ = when (daysAway) {
        0 -> "hoje"
        1 -> "amanhã"
        in 2..Int.MAX_VALUE -> weekdayResolver(nextDay.dateIso) ?: "em $daysAway dias"
        else -> "em breve"
    }
    val title = nextDay.first.title
    var line = "$when_, ${nextDay.first.startTime} · $title"
    val room = nextDay.first.room
    if (!room.isNullOrEmpty()) line += " — $room"
    return line
}

// Calendar-day distance between two `YYYY-MM-DD` strings, signed so a future
// `to` returns positive. Used to re-derive `daysAway` at render time from
// `nowDateIso` and the snapshot's `nextDay.dateIso` so the dayDone / upcoming
// branch picks the right state even when the snapshot is stale across a day
// boundary. Returns null if either input doesn't parse.
internal fun daysBetweenIsoDays(from: String, to: String): Int? {
    val a = parseIsoDay(from) ?: return null
    val b = parseIsoDay(to) ?: return null
    val diffMs = b.timeInMillis - a.timeInMillis
    // Use the per-day midpoint rounding to absorb DST transitions without
    // ever crossing a calendar-day boundary.
    return ((diffMs + (if (diffMs >= 0) 43_200_000L else -43_200_000L)) / 86_400_000L).toInt()
}

private fun parseIsoDay(iso: String): Calendar? {
    val parts = iso.split("-")
    if (parts.size != 3) return null
    val year = parts[0].toIntOrNull() ?: return null
    val month = parts[1].toIntOrNull() ?: return null
    val day = parts[2].toIntOrNull() ?: return null
    return Calendar.getInstance().apply {
        clear()
        set(year, month - 1, day)
    }
}

internal fun parseHhMm(value: String?): Int? {
    if (value.isNullOrBlank() || value.length < 4) return null
    val parts = value.split(":")
    if (parts.size < 2) return null
    val h = parts[0].toIntOrNull() ?: return null
    val m = parts[1].toIntOrNull() ?: return null
    return h * 60 + m
}

// Approximation of the design's `shortCode`. Upstream doesn't carry one, so
// we take the first whitespace-separated token of the discipline code, capped
// at 6 chars. Mirrors iOS's `shortCode`.
private fun shortCode(code: String): String {
    val first = code.split(" ").firstOrNull().orEmpty()
    return first.take(6)
}

// First clause before " — " / " – " / " - " / ":" if present, capped at 14
// chars. Upstream titles are full discipline names; the design shows a
// compact form. Mirrors iOS's `shortTitle`.
private fun shortTitle(title: String): String {
    val separators = listOf(" — ", " – ", " - ", ": ")
    var base = title
    for (sep in separators) {
        val idx = base.indexOf(sep)
        if (idx >= 0) base = base.substring(0, idx)
    }
    return if (base.length <= 14) base else base.take(13) + "…"
}

// 10-slot subject palette — same hex values + djb2 hashing as iOS
// `WidgetColor.subjectHex` and Android `ColorFor.discipline`. Carried here
// instead of consumed from the design system because Glance code is read
// outside `MaterialTheme` (which the design-system palette requires).
//
// Dark variants mirror `MelonPaletteColors` lifted-for-dark hexes
// (`apps/android/design-system/.../Color.kt`). The light slots are dim
// against the cool/dark plum mesh — particularly plum and indigo, which
// effectively disappear on dark surfaces — so we resolve to the lifted
// variant whenever the widget renders against the dark theme.
internal object SubjectPalette {
    private val light = longArrayOf(
        0xFFE85D4E, // coral
        0xFFF4A23C, // amber
        0xFFB23A7A, // magenta
        0xFF3B9EAE, // teal
        0xFF2D1B4E, // plum
        0xFFC64A6D, // rose
        0xFF3C7DC9, // sky
        0xFF2E8B5C, // emerald
        0xFF4A5FB8, // indigo
        0xFFA0741F, // mustard
    )

    private val dark = longArrayOf(
        0xFFF27E6E, // coral
        0xFFF4A23C, // amber
        0xFFD46299, // magenta
        0xFF5BB8C6, // teal
        0xFFB39DDB, // plum (lifted — was 0x2D1B4E, invisible on plum surface)
        0xFFE88AA5, // rose
        0xFF79AEE8, // sky
        0xFF5FC48E, // emerald
        0xFF8A9EE8, // indigo
        0xFFD4A84C, // mustard
    )

    fun argb(code: String, isDark: Boolean = false): Int {
        val palette = if (isDark) dark else light
        val bucket = (stableHash(code).let { if (it < 0) -it else it }) % palette.size
        return palette[bucket].toInt()
    }

    private fun stableHash(s: String): Int {
        var h = 5381
        for (c in s) h = ((h shl 5) + h) + c.code
        return h
    }
}

// Localized weekday name in pt-BR ("segunda", "terça", …). Used by the
// snapshot renderer when the next class is more than one day out so the
// dayDone copy reads naturally.
internal fun weekdayLabelForIsoDay(isoDay: String): String? = runCatching {
    val parts = isoDay.split("-")
    if (parts.size != 3) return null
    val year = parts[0].toInt()
    val month = parts[1].toInt()
    val day = parts[2].toInt()
    val cal = java.util.Calendar.getInstance(Locale.Builder().setLanguage("pt").setRegion("BR").build())
    cal.set(year, month - 1, day)
    val df = java.text.SimpleDateFormat("EEEE", Locale.Builder().setLanguage("pt").setRegion("BR").build())
    df.format(cal.time)
        .replace("-feira", "")
        .lowercase(Locale.Builder().setLanguage("pt").setRegion("BR").build())
}.getOrNull()
